### DISCUSSION TITLE
QUESTION:  Design Decision - ColumnDefinition.Width/MinWidth/MaxWidth StyledProps

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7711] [Q&A] [answered]
As I'm delving daily into Avalonia UI code, I am falling in love with this code base.

But, as you might expect, I also find occasional things that make me wonder, such as the choice to make the core properties of ColumnDefinition as StyledProperties (not Direct).

The RAM size of 3 "GridLength" structs should be about 32 bytes on a 64-bit OS.
     1. double x 3:  8 bytes - per double  x 3
     2. GridUnityType x3 : 1 byte - per "GridUnitType" enum value = 3 bytes ==> rounds up to 8.

**So 32 bytes total**, if all were DirectProperties (not requiring a Dictionary lookup).   And the benefit is that you only have one object on the heap (the ColumnDef itself), and property access is near instant (5 nsec = 200 million per sec).  **So you use less RAM, with fewer GC objects to track/clean, and it runs faster.**

But as the code is written...
Every ColumnDefinition has at least ONE property, which means that for each ColumnDef, you MUST create the "ValueStore" and associated objects on the heap, consuming 100 bytes+ total (including .NET header records/etc for heap), and adding several more objects for the GC to track/collect)....   

And so the current design decision seems to cost us: (1) More RAM, (2) more objects for GC to track/clean, and (3) worse performance.

**Am I missing something?  Was this a good design choice?**

My concern is that this type of choice may have been made uniformly throughout Avalonia source code (which would contribute to performance issues, while increasing object counts and RAM usage).

=
I truly appreciate the entire AvaloniaUI team.   You guys are making the world a better place, one SLOC at a time!  And I am thankful for all the free code you have provided here for us.   You guys are amazing.

My only motivation here is to make an excellent thing, even more awesome.  Avalonia should rule-the-world.

**Comment by amwx** [comment, marked as answer]
`StyledProperty` is basically the direct equivalent of WPF/UWP's `DependencyProperty` and works with the Animation and Styling systems. 95% of the time this is what you want to use. I had a similar thinking to you on this at one point, "oh direct property avoids the memory overhead of the ValueStore, it's better...". Not really. `DirectProperty` should generally be used for readonly properties, where the property will never need styling/animation support, or collections (and this one should be fixed at some point in the future). 

Grid Column/Row definitions should be styleable components. That said, they currently are not - at least from Xaml. (#6276 #4397) I know you're using more of the Declarative stuff, so the xaml stuff is probably less important to you, but the idea is still the same

**Reply by najak3d** [reply]
Thank you for your answer!

That seems a bit unfortunate that DirectProperty cannot be Styled.   Whenever Avalonia suffers performance issues, these types of things are often the cause (since you can't get away without doing the extensive iterative Measure/Arrange calls, if you could speed up each pass, then you could cut the time cost to under half, etc).   So if there were a way to make DirectProperties non-deficient, then this could:
1. Speed up Avalonia by 2x+ (???)
2. Reduce Object generation, alleviating GC workload.

As we are creating a Game Editor and Game GUI concept, it won't sell well if Avalonia is: 
1. Causing slowdowns/lag/big-pauses
2. Or occasional 0.1 sec+ Hiccups from GC collection.  (causing 0.1 second frame rate blips)

When writing .NET games, it's vital to keep object/reference counts low, to prevent the GC 0 Hiccups, that may result otherwise.  At this point, I don't foresee Avalonia introducing any significant risk for object counts. 

We are also targeting games that run on mobile - where memory is more scan, and processing power is considerably lower than that of a PC.   And so I will continue to scrutinize the Avalonia code, for efficiency/performance concerns.

**Reply by amwx** [reply]
Also, `Grid` by nature is kind of a heavy control and if possible (and sometimes its not) it may be better to use one of the existing Panels (like DockPanel) or write your own. It comes from WPF so it's old code that really needs a refactor - there's just nothing yet to base it on.

> And so I will continue to scrutinize the Avalonia code, for efficiency/performance concerns.

Good. Increasing performance/efficiency benefits everyone!
